fpga-arch/bsv/SWD.bsv

```diff
@@ -0,0 +1,196 @@
+    (* prefix="" *) method Action write((* port="write" *) Bit#(1) b);
+    (* prefix="" *) method Action read((* port="read" *) Bit#(1) b);
+
+    // register selection
+    (* prefix="" *) method Action en_rdata((* port="en_rdata" *) Bit#(1) en);
+    (* prefix="" *) method Action en_wdata((* port="en_wdata" *) Bit#(1) en);
+    (* prefix="" *) method Action en_cmd((* port="en_cmd" *) Bit#(1) en);
+    (* prefix="" *) method Action en_status((* port="en_status" *) Bit#(1) en);
+
+    // readable register out
+    (* prefix="" *) method Bit#(8) reg_rdata;
+    (* prefix="" *) method Bit#(8) reg_status;
+endinterface
+
+interface ScaffoldSWDModule;
+    (* prefix="" *) interface ScaffoldBus bus;
+    (* prefix="" *) interface SWDControllerPins pins;
+    (* always_ready, prefix="" *) method Bit#(1) trigger;
+endinterface
+
+typedef struct {
+    Bit#(1) reset;
+    Bit#(1) trigger;
+    Bit#(2) reserved;
+    Bit#(1) apndp;
+    Bit#(1) rnw;
+    Bit#(2) addr;
+} Cmd deriving (Eq, Bits);
+
+typedef enum {
+    IDLE,
+    RESET,
+    RW
+} State deriving (Eq, Bits);
+
+(* synthesize *)
+module swd_module (ScaffoldSWDModule);
+    SWDController#(100) swd_controller <- mkSWDController();
+
+    Wire#(Bit#(16)) bus_address <- mkDWire(0);
+    Wire#(Bit#(8)) bus_write_data <- mkDWire(0);
+    Wire#(Bit#(1)) bus_write <- mkDWire(0);
+    Wire#(Bit#(1)) bus_read <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_rdata <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_wdata <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_cmd <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_status <- mkDWire(0);
+
+    Reg#(Bit#(8)) bus_reg_rdata <- mkRegA(0);
+    Reg#(Bit#(8)) bus_reg_status <- mkRegA(0);
+
+    PulseWire trig <- mkPulseWire();
+
+    Reg#(Vector#(4, Bit#(8))) rdata <- mkRegA(unpack(0));
+    Reg#(Status) status <- mkRegA(unpack(0));
+    Reg#(Vector#(4, Bit#(8))) wdata <- mkRegA(unpack(0));
+    Reg#(Maybe#(Cmd)) cmd <- mkRegA(tagged Invalid);
+
+    Reg#(Bool) ready <- mkRegA(False);
+    Reg#(State) state <- mkRegA(IDLE);
+
+    rule do_bus_read_rdata ((bus_read == 1) && (bus_en_rdata == 1) && (state != RW));
+        bus_reg_rdata <= rdata[0];
+        rdata <= shiftInAtN(rdata, 0);
+    endrule
+
+    rule do_bus_read_status ((bus_read == 1) && (bus_en_status == 1));
+        bus_reg_status <= {pack(state == IDLE), 5'b0, pack(status)};
+    endrule
+
+    rule do_bus_write ((bus_write == 1) && (state == IDLE));
+        if (bus_en_wdata == 1) begin
+            wdata <= shiftInAt0(wdata, bus_write_data);
+        end
+        
+        if (bus_en_cmd == 1) begin
+            cmd <= tagged Valid(unpack(bus_write_data));
+        end
+    endrule
+
+    rule do_ready;
+        ready <= swd_controller.ready;
+    endrule
+
+    // ONLY do something if there is a valid command registered, and if
+    // the user is not currently writing to some register.
+    rule do_idle ((state == IDLE) && (bus_write == 0) && isValid(cmd));
```

Comment:
I think `bus_write == 0` should not be in the condition here. As long as `cmd` is valid, the SWD peripheral can start working.
In particular, the host may be talking with another Scaffold peripheral at the same time, and this shall not prevent the SWD from working.